proposals/list-patterns.md

```diff
@@ -0,0 +1,122 @@
+# List patterns
+
+## Summary
+
+Lets you to match an array or a list with a sequence of patterns e.g. `array is {1, 2, 3}` will match an integer array of the length three with 1, 2, 3 as its elements, respectively.
+
+## Detailed Design
+
+The pattern syntax is modified as follow:
+
+```antlr
+recursive_pattern
+  : type? positional_pattern_clause? length_pattern_clause? property_or_list_pattern_clause? simple_designation?
+
+length_pattern_clause
+  : '[' pattern ']'
+
+property_or_list_pattern_clause
+  : list_pattern_clause
+  | property_pattern_clause
+
+property_pattern_clause
+  : '{' (subpattern (',' subpattern)* ','?)? '}'
+  
+list_pattern_clause
+  : '{' pattern (',' pattern)* ','? '}'
+
+slice_pattern
+  : '..' negated_pattern?
+	
+primary_pattern
+  : recursive_pattern
+  | slice_pattern
+
+```
+There are two new additions to the *recursive pattern* syntax as well as a *slice pattern*: 
+
+- The *list pattern clause* is used to match elements and the *length pattern clause* is used to match the length.
+- A _slice pattern_ is only permitted once and only directly in a *list pattern* and discards _**zero or more**_ elements.
+
+
+Due to the ambiguity with *property pattern clause*, the *list pattern clause* cannot be empty and a *length pattern* should be used. 
```

Comment:
```suggestion
Due to the ambiguity with *property pattern clause*, the *list\_pattern\_clause* cannot be empty and a *length\_pattern* should be used. 
```

Reply:
Similar changes are needed throughout the rest. Not going to go through and suggest for them all.